packages/indexer/tests/db_test.py

```diff
@@ -103,6 +170,75 @@ async def test_upsert_raw_blocks(
         assert res["block_hash"] == block["block_id"]["hash"]
         assert res["proposer_address"] == block["block"]["header"]["proposer_address"]
 
-    # assert await get_missing_blocks(mock_pool, mock_session, mock_sem, mock_chain) == []
+    await drop_tables(mock_db)
+
+
+@pytest.mark.asyncio
+async def test_upsert_raw_txs(
+    mock_db: Database,
+    mock_listener: asyncpg_listen.NotificationListener,
+    mock_chain: CosmosChain,
+    mock_client: aiohttp.ClientSession,
+    mock_semaphore: asyncio.Semaphore,
+    raw_txs: pd.DataFrame,
+    mock_notification_handler: DbNotificationHandler,
+    mocker,
+):
+    await create_tables(mock_db)
+
+    listener_task = asyncio.create_task(
+        mock_listener.run(
+            {"txs_to_logs": mock_notification_handler.process_tx_notifications},
+            policy=asyncpg_listen.ListenPolicy.ALL,
+        )
+    )
+
+    flattened_raw_txs = []
+    for _chain_id, height, txs, _parsed_at in raw_txs[:500].itertuples(index=False):
+        flattened_raw_txs.extend(txs)
+        await upsert_raw_txs(mock_db, {int(height): txs}, mock_chain.chain_id)
+
+    # valid block data
+    for _chain_id, height, txs, _parsed_at in raw_txs[500:].itertuples(index=False):
+        flattened_raw_txs.extend(txs)
+        mocker.patch(
+            "indexer.process.CosmosChain.get_block_txs",
+            return_value={"tx_responses": txs},
+        )
+
+        assert True == await process_tx(
+            height, mock_chain, mock_db, mock_client, mock_semaphore
+        )
+
+    # invalid tx data
+    mocker.patch("indexer.process.CosmosChain.get_block_txs", return_value=None)
+    assert False == await process_tx(
+        height, mock_chain, mock_db, mock_client, mock_semaphore
+    )
+
+    async with mock_db.pool.acquire() as conn:
+        raw_results = await conn.fetch(
+            f"""
+            select * from {mock_db.schema}.raw_txs
+            """
+        )
+
+        results = await conn.fetch(f"select * from {mock_db.schema}.txs")
+
+    assert len(results) == len(flattened_raw_txs)
+
+    while len(mock_notification_handler.notifications) < len(flattened_raw_txs):
```

Comment:
wait to see if number of notifications match number of txs, then close notification handler